packages/aws-cdk/lib/util/asset-publishing.ts

```diff
@@ -66,16 +68,32 @@ class PublishingAws implements cdk_assets.IAws {
   /**
    * Get an SDK appropriate for the given client options
    */
-  private sdk(options: cdk_assets.ClientOptions): Promise<ISDK> {
+  private async sdk(options: cdk_assets.ClientOptions): Promise<ISDK> {
+    // if this was `JSON.stringify(options)`, extra fields might
+    // accidentally be included in the cache key
+    const cacheKey = JSON.stringify({
+      region: options.region,
+      assumeRuleArn: options.assumeRoleArn,
+      assumeRoleExternalId: options.assumeRoleExternalId,
+    });
+
+    const maybeSdk = this.sdkCache.get(cacheKey);
+    if (maybeSdk) {
+      return Promise.resolve(maybeSdk);
```

Comment:
The function is already async

```suggestion
      return maybeSdk;
```